impeller/display_list/dl_dispatcher.cc

```diff
@@ -622,23 +622,11 @@ void DlDispatcherBase::saveLayer(const SkRect& bounds,
   auto promise = options.content_is_clipped()
                      ? ContentBoundsPromise::kMayClipContents
                      : ContentBoundsPromise::kContainsContents;
-  std::optional<Rect> impeller_bounds;
-  if (!options.content_is_unbounded() || options.bounds_from_caller()) {
-    impeller_bounds = skia_conversions::ToRect(bounds);
-  }
-
-  // Empty bounds on a save layer that contains a BDF or destructive blend
-  // should be treated as unbounded. All other empty bounds can be skipped.
-  if (impeller_bounds.has_value() && impeller_bounds->IsEmpty() &&
-      (backdrop != nullptr ||
-       Entity::IsBlendModeDestructive(paint.blend_mode))) {
-    impeller_bounds = std::nullopt;
-  }
-
+  std::optional<Rect> impeller_bounds = skia_conversions::ToRect(bounds);
   GetCanvas().SaveLayer(paint, impeller_bounds, ToImageFilter(backdrop),
                         promise, total_content_depth,
                         options.can_distribute_opacity(),
-                        options.bounds_from_caller());
+                        options.bounds_from_caller() && !backdrop);
```

Comment:
The "bounds from caller" hint isn't useful here AFAICT. If it matters then the bounds are being used wrong. I use it internally in the Builder to know if the recorded op has bounds from the caller or "dummy" bounds. I then combine the 2 in the restore call. At that point "bounds from caller" has essentially completed its entire "raison d'être" and is obsolete. I leave it in there because it isn't really useful to delete it - the receiver's shouldn't have any use for it.

In *all* cases the bounds are the bounds of the *content* of the saveLayer. How they were determined may be interesting, but not really relevant to the fact that they are accurately the bounds of what the code between the saveLayer and its corresponding restore draw, potentially clipped by the user supplied bounds (in which case the "clipped" flag will be set in the options.

When determining how much to allocate, you would determine:

- If there is a BDF, then allocate enough to map to the current clip bounds.
- If the BDF somehow were to indicate that it isn't going to draw into those entire bounds, you can take that into account, but then you would also need to factor in the contents.
- If you don't query the BDF about its output intentions, or if the BDF does indicate that it will draw to the entire target area, then you are done and the clip bounds are the answer.
- If there is no BDF, then the RT needs to cover the indicated bounds from the saveLayer call transformed by the CTM.

Reply:
Another way to look at this, the saveLayer will cause 2 kinds of rendering:

- If there is a BDF, then the BDF will filter everything underneath the saveLayer and come up with some output pixels. That will produce output relative as per the BDF and the CTM of the saveLayer and the input of the BDF is the pixels from the parent RT which are in device pixels. That's one source of rendering.
- Whether there is a BDF, the indicated saveLayer bounds will represent how much you need to hold the content that was between saveLayer and Restore. These bounds are relative to the CTM.
- Combine those 2, and clip to the current clip bounds and that represents the pixels you want to capture for the eventual blit to the parent RT.

The saveLayer bounds may be irrelevant if the BDF says "I plan to render to the entire surface" then what does it matter what the saveLayer content does? But, if we ask the BDF "what's your output if you transform these pixels expressed in the device coordinate system by the BDF acting in the CTM?" then the 2 bounds may need to be combined.